orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -59,11 +62,56 @@ void FindLeafNodes(Graph& graph, std::vector<NodeArg*>& input_args) {
       }
     }
     if (!find_consumer_nodes && outputs.size() > 0) {
-      input_args.push_back(outputs[0]);
+      leaf_node_args.push_back(outputs[0]);
     }
   }
+
+  return leaf_node_args;
 };
 
+// This function converts tensor NodeArg to a boolean scalar so that last
+// backward RecordEvent doesn't block the early release of large gradient
+// tensors. If we connect gradient tensors directly to that RecordEvent,
+// we need a memory block (as large as a whole model) to store gradient
+// for each trainable tensor until the end of backward pass.
+//
+// The newly created bollean scalar may be appended to signal_args. If
```

Comment:
>bollean [](start = 21, length = 7)

nit: typo..

Reply:
XD. Thanks.

---
In reply to: [439676635](https://github.com/microsoft/onnxruntime/pull/4204#discussion_r439676635) [](ancestors = 439676635)